test/ng/ngPropSpec.js

```diff
@@ -0,0 +1,775 @@
+      }));
+
+      it('should handle custom $sce objects', function() {
+        function MySafeHtml(val) { this.val = val; }
+
+        module(function($provide) {
+          $provide.decorator('$sce', function($delegate) {
+            $delegate.trustAsHtml = function(html) { return new MySafeHtml(html); };
+            $delegate.getTrusted = function(type, mySafeHtml) { return mySafeHtml && mySafeHtml.val; };
+            $delegate.valueOf = function(v) { return v instanceof MySafeHtml ? v.val : v; };
+            return $delegate;
+          });
+        });
+
+        inject(function($rootScope, $compile, $sce) {
+          // Ref: https://github.com/angular/angular.js/issues/14526
+          // Previous code used toString for change detection, which fails for custom objects
+          // that don't override toString.
+          var element = $compile('<div ng-prop-inner_h_t_m_l="getHtml()"></div>')($rootScope);
+          var html = 'hello';
+          $rootScope.getHtml = function() { return $sce.trustAsHtml(html); };
+          $rootScope.$digest();
+          expect(lowercase(element.html())).toEqual('hello');
+          html = 'goodbye';
+          $rootScope.$digest();
+          expect(lowercase(element.html())).toEqual('goodbye');
+        });
+      });
+
+      describe('when $sanitize is available', function() {
+        beforeEach(function() { module('ngSanitize'); });
+
+        it('should sanitize untrusted html', inject(function($rootScope, $compile) {
+          var element = $compile('<div ng-prop-inner_h_t_m_l="html"></div>')($rootScope);
+          $rootScope.html = '<div onclick="">hello</div>';
+          $rootScope.$digest();
+          expect(lowercase(element.html())).toEqual('<div>hello</div>');
+        }));
+      });
+    });
+
+  });
+
+  describe('*[style]', function() {
+    it('should NOT set style for untrusted values', inject(function($rootScope, $compile) {
+      var element = $compile('<div ng-prop-style="style"></div>')($rootScope);
+      $rootScope.style = 'background: red';
+      expect(function() { $rootScope.$digest(); }).toThrowMinErr('$sce', 'unsafe', 'Attempting to use an unsafe value in a safe context.');
+    }));
+
+    it('should NOT set style for wrongly typed values', inject(function($rootScope, $compile, $sce) {
+      var element = $compile('<div ng-prop-style="style"></div>')($rootScope);
+      $rootScope.style = $sce.trustAsHtml('background: red');
+      expect(function() { $rootScope.$digest(); }).toThrowMinErr('$sce', 'unsafe', 'Attempting to use an unsafe value in a safe context.');
+    }));
+
+    it('should set style for trusted values', inject(function($rootScope, $compile, $sce) {
+      var element = $compile('<div ng-prop-style="style"></div>')($rootScope);
+      $rootScope.style = $sce.trustAsCss('background: red');
+      $rootScope.$digest();
+      expect(lowercase(element.css('background-color'))).toEqual('red');
```

Comment:
A lot of browsers (which allow assigning to `HTMLElement.style`) seem to return `rgb(255,0,0)` instead of `red` here. I'm going to change this to a property that won't have normalization issues like that.

Reply:
See 125c7b138374409ca96fe51bd125219a518d8c1f